src/Java.Interop.Tools.Expressions/Java.Interop.Tools.Expressions/ExpressionAssemblyBuilder.cs

```diff
@@ -0,0 +1,452 @@
+		// il.Emit (OpCodes.Stloc_0);
+
+		var JniNativeMethodRegistration_ctor    = typeof (JniNativeMethodRegistration).GetConstructor (new [] { typeof (string), typeof (string), typeof (Delegate) });
+		var jnmr_ctor                           = DeclaringAssemblyDefinition.MainModule.ImportReference (JniNativeMethodRegistration_ctor);
+
+		for (int i = 0; i < methods.Count; i++) {
+			var delegateCtor = GetMarshalMethodDelegateCtor (methods [i].MarshalMethodDefinition);
+
+			// il.Emit (OpCodes.Ldloc_0);      // args
+			il.Emit (OpCodes.Dup);      // args
+			il.Emit (OpCodes.Ldc_I4, i);    // index of `args` to set
+
+			// new JniNativeMethodRegistration (JniName, JniSignature, new _JniMarshal_PP… (MarshalMethodDefinition))
+			il.Emit (OpCodes.Ldstr,         methods [i].JniName);
+			il.Emit (OpCodes.Ldstr,         methods [i].JniSignature);
+			il.Emit (OpCodes.Ldnull);
+			il.Emit (OpCodes.Ldftn,         methods [i].MarshalMethodDefinition);
+			il.Emit (OpCodes.Newobj,        delegateCtor);
+			il.Emit (OpCodes.Newobj,        jnmr_ctor);
+
+			il.Emit (OpCodes.Stelem_Any,	arrayType.GetElementType ());   // args [i] = new JniNativeMethodRegistration (…)
+		}
+
+		il.Emit (OpCodes.Stloc_0);
+
+		Action<IEnumerable<JniNativeMethodRegistration>> addRegistrations = new JniNativeMethodRegistrationArguments ().AddRegistrations;
+		il.Emit (OpCodes.Ldarga_S, args);
+		il.Emit (OpCodes.Ldloc_0);
+		il.Emit (OpCodes.Call, DeclaringAssemblyDefinition.MainModule.ImportReference (addRegistrations.Method));
+		il.Emit (OpCodes.Ret);
+
+
+		return registrations;
+	}
+
+	void EmitConsoleWriteLine (ILProcessor il, string message)
+	{
+		Action<string> cwl = Console.WriteLine;
+		il.Emit (OpCodes.Ldstr, message);
+		il.Emit (OpCodes.Call,  DeclaringAssemblyDefinition.MainModule.ImportReference (cwl.Method));
+	}
+
+	// Keep in sync w/ MarshalMemberBuilder.GetMarshalerType()
+	MethodReference GetMarshalMethodDelegateCtor (MethodDefinition method)
+	{
+		// Too many parameters; does a `_JniMarshal_*` type exist in the type's declaring assembly?
+		var delegateName    = GetMarshalMethodDelegateName (method.Parameters, method.ReturnType);
+
+		var delegateDef = DeclaringAssemblyDefinition.MainModule.GetType (delegateName.ToString ());
+		if (delegateDef == null) {
+			delegateDef     = CreateMarshalMethodDelegateType (delegateName, method.Parameters, method.ReturnType);
+			DeclaringAssemblyDefinition.MainModule.Types.Add (delegateDef);
+		}
+		return delegateDef.Methods.First (m => m.Name == ".ctor");
+	}
+
+	string GetMarshalMethodDelegateName (IList<ParameterDefinition> parameters, TypeReference returnType)
+	{
+		// Too many parameters; does a `_JniMarshal_*` type exist in the type's declaring assembly?
+		var delegateName    = new StringBuilder ();
+		delegateName.Append ("_JniMarshal_PP");
+
+		for (int i = 2; i < parameters.Count; i++) {
+			delegateName.Append (GetJniMarshalDelegateParameterIdentifier (parameters [i].ParameterType));
```

Comment:
I might have missed it, but shouldn't it handle arrays too?

Reply:
Arrays will hit the default case of `L`, which is "any pointer type" at the ABI boundary.  This is akin to xamarin/xamarin-android@baa5a739b957d724df3a33f001e4e1cde914cee5, which doesn't need to care about how many `[` are encountered, as no matter how deep the array is, it's still just a pointer.